File: 20-valid-parentheses/valid-parentheses.java
import java.util.Stack;

class Solution {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                // Opening bracket - push to stack
                stack.push(c);
            } else {
                // Closing bracket - check if stack is empty
                if (stack.isEmpty()) {
                    return false;
                }
                
                // Get the top element and check if it matches
                char top = stack.pop();
                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return false;
                }
            }
        }
        
        // If stack is empty, all brackets were properly matched
        return stack.isEmpty();
    }
}

File: 20-valid-parentheses/valid-parentheses.php
class Solution {

    /**
     * @param String $s
     * @return Boolean
     */
    function isValid($s) {
        $stack = [];
        $mapping = [
            ')' => '(',
            '}' => '{',
            ']' => '['
        ];
        
        for ($i = 0; $i < strlen($s); $i++) {
            $char = $s[$i];
            
            // If it's a closing bracket
            if (array_key_exists($char, $mapping)) {
                // Pop the top element if stack is not empty, otherwise use a dummy value
                $topElement = empty($stack) ? '#' : array_pop($stack);
                
                // If the mapping doesn't match, return false
                if ($topElement != $mapping[$char]) {
                    return false;
                }
            } else {
                // If it's an opening bracket, push to stack
                array_push($stack, $char);
            }
        }
        
        // If stack is empty, all brackets were properly closed
        return empty($stack);
    }
}